qiskit/mapper/_layout.py

```diff
@@ -0,0 +1,145 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2018, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+
+"""
+A two-ways dict that represent a layout.
+
+Layout is the relation between logical (qu)bits and physical wires.
+Logical (qu)bits are tuples (eg, `('qr',2)`.
+Physical wires are numbers.
+"""
+
+from qiskit import QISKitError
+
+
+class Layout(dict):
+    """ Two-ways dict to represent a Layout."""
+
+    def __init__(self, input_=None):
+        dict.__init__(self)
+        if isinstance(input_, dict):
+            self.from_dict(input_)
+        if isinstance(input_, list):
+            self.from_list(input_)
+
+    def from_dict(self, input_dict):
+        """
+        Pullulates a Layout from a dictionary.
+
+        Args:
+            input_dict (dict): For example,
+            {('qr', 0): 0, ('qr', 1): 1, ('qr', 2): 2}
+        """
+        for key, value in input_dict.items():
+            self[key] = value
+
+    def from_list(self, input_list):
+        """
+        Pullulates a Layout from a list.
+
+        Args:
+            input_list (list): For example,
+            [('qr', 0), None, ('qr', 2), ('qr', 3)]
+        """
+        for key, value in enumerate(input_list):
+            self[key] = value
+
+    def __getitem__(self, item):
+        if item is None:
+            return None
+        if isinstance(item, int) and item < len(self) and item not in self:
+            return None
+        return dict.__getitem__(self, item)
+
+    def __setitem__(self, key, value):
+        if key in self:
+            del self[key]
+        if value in self:
+            del self[value]
+        dict.__setitem__(self, key, value)
+        dict.__setitem__(self, value, key)
+
+    def __delitem__(self, key):
+        dict.__delitem__(self, self[key])
+        dict.__delitem__(self, key)
+
+    def __len__(self):
+        return max([key for key in self.keys() if isinstance(key, int)], default=-1) + 1
+
+    def add(self, logical, physical=None):
+        """
+        Adds a map element between `logical` and `physical`. If `physical`
+        is not defined, `logical will be mapped to a new wire (extending
+        the length of the layout)
+        Args:
+            logical (tuple): A logical (qu)bit. For example, ('qr',2).
+            physical (int): A physical wire. For example, 3.
+        """
+        if physical is None:
+            physical = len(self)
+        self[logical] = physical
+
+    def length(self, amount_of_wires):
```

Comment:
should this be `set_length` or `extend` or somthing? `length` sounds like it would just return a length.

Reply:
changed to `set_length` in 77e1c73